### PR TITLE
Fix buffer dtype in ` FSDP.state_dict()` when using mixed precision

### LINKED ISSUE #704: FSDP state_dict returned data type when using mixed precision
The docstring for `FSDP.state_dict()` says that returned parameters will be in FP32: https://github.com/facebookresearch/fairscale/blob/b84b9146aca8fd575622462c1b311243e40b115c/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L647-L648.

But this is not the case when `mixed_precision=True`. So what's the best way to get the FP32 parameters when using mixed precision?

### PR DESCRIPTION
**Pull request by epwalsh**
## What does this PR do?

Fixes #704.

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [x] N/A
- [x] Did you write any new necessary tests?
  - [ ] N/A
- [x] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by min-xu-ai** [comment]
Thanks! Cc @sshleifer

**Comment by epwalsh** [comment]
@min-xu-ai, @sshleifer, I think I came up with a fix.

**Comment by min-xu-ai** [comment]
Wow. This is great! I think there might some python formatting or mypy errros that CI is detecting. Once you fixed those, I can merge it.

**Comment by sshleifer** [comment]
LGTM thanks!

**Comment by epwalsh** [comment]
@min-xu-ai I've ran `flake8`, `black`, and `mypy` all locally according to the `CONTRIBUTING.md` document. I'm not seeing any errors. And the only error I'm seeing in CI is an installation issue that seems unrelated.

**Comment by min-xu-ai** [comment]
You are right. It is unrelated. Let us first fix it on the master and then you can merge from the master and we can merge this in. Thanks again!

**Comment by epwalsh** [comment]
Sounds good, thanks @min-xu-ai!

**Comment by min-xu-ai** [comment]
> Sounds good, thanks @min-xu-ai!

It is fixed now: https://github.com/facebookresearch/fairscale/pull/706

Can you merge master and push to this PR? CI should run again with that fix. It should pass and then I can merge.

File: .circleci/config.yml [hunk 1]
```diff
@@ -114,7 +114,7 @@ install_dep_190: &install_dep_190
         # check if we have restored venv cache (/home/circleci/venv) correctly, if so, just skip
         if [ -f /home/circleci/venv/check_version.py ]; then python /home/circleci/venv/check_version.py torch eq 1.8 && exit 0; fi
         # start installing
-        pip install --pre --progress-bar off torch==1.9.0.dev20210330+cu101 torchvision==0.10.0.dev20210330+cu101 -f https://download.pytorch.org/whl/nightly/cu101/torch_nightly.html
+        pip install --pre --progress-bar off torch==1.9.0.dev20210415+cu101 torchvision==0.10.0.dev20210415+cu101 -f https://download.pytorch.org/whl/nightly/cu101/torch_nightly.html
```
Comment (min-xu-ai) [inline .circleci/config.yml:117]: this is strange. it shouldn't show up in your PR. can you make sure you merge the master, not cherry pick that gives the commit already in master a different git sha1 hash?

Comment (min-xu-ai) [reply .circleci/config.yml:117]: never mind, I think the commit is correct with the same sha1 hash as the master. the github's way of doing it confused me. Sorry about that.

**Comment by min-xu-ai** [comment]
I think this is all good now. Will merge once the CI passes.

File: .circleci/config.yml (hunk 1 above)
Comment (epwalsh) [reply .circleci/config.yml]: That was weird, it's fixed now though.

**Pull request merged**